pkg/apis/core/validation/validation.go

```diff
@@ -2020,15 +2020,20 @@ func ValidatePersistentVolumeStatusUpdate(newPv, oldPv *core.PersistentVolume) f
 	return allErrs
 }
 
-// PersistentVolumeClaimSpecValidationOptions contains the different settings for PersistentVolumeClaim validation
 type PersistentVolumeClaimSpecValidationOptions struct {
 	// Allow spec to contain the "ReadWiteOncePod" access mode
 	AllowReadWriteOncePod bool
+	// Allow pvc expansion after PVC is created and bound to a PV
+	EnableExpansion bool
+	// Allow users to recover from previously failing expansion operation
+	EnableRecoverFromExpansionFailure bool
 }
 
 func ValidationOptionsForPersistentVolumeClaim(pvc, oldPvc *core.PersistentVolumeClaim) PersistentVolumeClaimSpecValidationOptions {
 	opts := PersistentVolumeClaimSpecValidationOptions{
-		AllowReadWriteOncePod: utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod),
+		AllowReadWriteOncePod:             utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod),
+		EnableExpansion:                   utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes),
+		EnableRecoverFromExpansionFailure: utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure),
```

Comment:
To support [ratcheting validation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field) with api server version skew, the n-1 version of the api server needs to be able to allow objects that that the n version of the api server allowed.

So in version n where the feature is turned on by default, the n-1 validation needs to be able to allow PVC objects where the size shrunk. So I think one more check may need to be added here to say "if the old object already allowed shrinking, then the EnableRecoverFromExpansionFailure option is true"

Reply:
Discussed this bit offline and I am slightly confused. The ratcheting validation mechanism ensures that API objects stored in api-server with feature gate enabled, remains valid when feature gate is disabled (on older api-server if I understood this right). 

One key critirea is: older api-server must explicitly reject unknown enum values with something like following:

```
 if f.RestartPolicy == RestartPolicyOnTuesday && !opts.AllowRestartPolicyOnTuesday {
    allErrs = append(allErrs, field.Invalid(field.NewPath("restartPolicy"), f.RestartPolicy, "only allowed if the Fro
```

But we aren't rejecting anything if feature-gate is disabled in `pvc.Status`. May be if we did - then we would need this mechanism, but since `status` field is basically not user-modifiable and validations ignore both `resizeStatus` and `allocatedResources` if feature-gate is disabled, we should be fine?

Reply:
Discussed offline. I misread the validation logic and I think the added validation logic works. 1.23 logic will allow a volume shrink from 1.24 to continue to exist. Can you add a test case for these scenarios? ie RecoverExpansion feature is off, but the old pvc object has the new fields, and old.spec.capacity is less than old.status.allocated?

Also, do we have a test case where: feature is disabled and someone tries to shrink?

Reply:
We do - https://github.com/kubernetes/kubernetes/pull/106154/files#diff-7ee023b4401073292170227985c8ac98015db0c1980657c277a7cea011c24830R2128